nestkernel/secondary_event.h

```diff
@@ -0,0 +1,464 @@
+/*
+ *  secondary_event.h
+ *
+ *  This file is part of NEST.
+ *
+ *  Copyright (C) 2004 The NEST Initiative
+ *
+ *  NEST is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef SECONDARY_EVENT_H
+#define SECONDARY_EVENT_H
+
+namespace nest
+{
+
+/**
+ * Base class of secondary events. Provides interface for
+ * serialization and deserialization. This event type may be
+ * used to transmit data on a regular basis
+ * Further information about secondary events and
+ * their usage with gap junctions can be found in
+ *
+ * Hahne, J., Helias, M., Kunkel, S., Igarashi, J.,
+ * Bolten, M., Frommer, A. and Diesmann, M.,
+ * A unified framework for spiking and gap-junction interactions
+ * in distributed neuronal network simulations,
+ * Front. Neuroinform. 9:22. (2015),
+ * doi: 10.3389/fninf.2015.00022
+ */
+class SecondaryEvent : public Event
+{
+
+public:
+  SecondaryEvent* clone() const override = 0;
+
+  virtual void add_syn_id( const synindex synid ) = 0;
+
+  virtual bool supports_syn_id( const synindex synid ) const = 0;
+
+  //! size of event in units of unsigned int
+  virtual size_t size() = 0;
+  virtual std::vector< unsigned int >::iterator& operator<<( std::vector< unsigned int >::iterator& pos ) = 0;
+  virtual std::vector< unsigned int >::iterator& operator>>( std::vector< unsigned int >::iterator& pos ) = 0;
+
+  virtual const std::vector< synindex >& get_supported_syn_ids() const = 0;
+
+  virtual void reset_supported_syn_ids() = 0;
+};
+
+/**
+ * This template function returns the number of uints covered by a variable of
+ * type T. This function is used to determine the storage demands for a
+ * variable of type T in the NEST communication buffer, which is of type
+ * std::vector<unsigned int>.
+ */
+template < typename T >
+size_t
+number_of_uints_covered()
```

Comment:
Maybe better 
```suggestion
size_in_uints()
```
And could this entire thing be turned into a constexpr, something like (not 100% sure the math works out)
```
{
  return ( sizeof( T ) + sizeof( unsigned int ) - 1 ) / sizeof( unsigned int );
}
```

Reply:
This PR is not about rewriting the code. Please file another PR if you have suggestions like these.